src/components/cuda_pw/linux-cuda.c

```diff
@@ -0,0 +1,377 @@
+
+    if (_cuda_pw_vector.cmp_info.initialized
+        && _cuda_pw_vector.cmp_info.disabled == PAPI_EDELAY_INIT
+    ) {
+        cuda_init_private();
+    }
+
+    _papi_hwi_unlock(COMPONENT_LOCK);
+}
+
+static int cuda_ntv_enum_events(unsigned int *event_code, int modifier)
+{
+    check_n_initialize();
+
+    (void) event_code;
+    (void) modifier;
+    return PAPI_ENOEVNT;  // Not implemented
+}
+
+static int cuda_ntv_name_to_code(const char *name, unsigned int *event_code)
+{
+    check_n_initialize();
+    int res;
+    event_rec_t *evt_rec;
+    res = find_event_name(&global_event_names, name, &evt_rec);
+    if (res == PAPI_OK) {
+        *event_code = evt_rec->evt_code;
+        return PAPI_OK;
+    }
+    else {
+        _papi_hwi_lock(COMPONENT_LOCK);
+        *event_code = global_event_names.count;
+        res = insert_event_record(&global_event_names, name, global_event_names.count, 0);
+        _papi_hwi_unlock(COMPONENT_LOCK);
+    }
+
+    return PAPI_OK;
+}
+
+static int cuda_ntv_code_to_name(unsigned int event_code, char *name, int len)
+{
+    check_n_initialize();
+    if (event_code >= global_event_names.count) {
+        return PAPI_EINVAL;
+    }
+
+    strncpy(name, global_event_names.evts[event_code].name, len);
+    return PAPI_OK;
+}
+
+static int cuda_ntv_code_to_descr(unsigned int event_code, char *descr, int len)
+{
+    check_n_initialize();
+    // strcpy(descr, metric_table->metrics[event_code].description);
+    (void) event_code; (void) descr; (void) len;
+    return PAPI_OK;
+}
+
+static int cuda_init_thread(hwd_context_t *ctx)
+{
+    (void) ctx;
+    return PAPI_OK;
+}
+
+static int cuda_shutdown_thread(hwd_context_t *ctx)
+{
+    (void) ctx;
+    return PAPI_OK;
+}
+
+static int cuda_init_control_state(hwd_control_state_t *ctl)
+{
+    COMPDBG("Entering.\n");
+    (void) ctl;
+    return PAPI_OK;
+}
+
+static int cuda_set_domain(hwd_control_state_t * ctrl, int domain)
+{
+    COMPDBG("Entering\n");
+    (void) ctrl;
+    if((PAPI_DOM_USER & domain) || (PAPI_DOM_KERNEL & domain) || (PAPI_DOM_OTHER & domain) || (PAPI_DOM_ALL & domain))
+        return (PAPI_OK);
+    else
+        return (PAPI_EINVAL);
+    return (PAPI_OK);
+}
+
+static int cuda_update_control_state(hwd_control_state_t *ctl,
+                                     NativeInfo_t *ntv_info,
+                                     int ntv_count, hwd_context_t *ctx
+) {
+    COMPDBG("Entering with events_count %d.\n", ntv_count);
+    check_n_initialize();
+    int i, res;
+    if (ntv_count == 0)
+        goto fn_exit;
+    struct cuda_ctl *control = (struct cuda_ctl *) ctl;
+    struct cuda_ctx *this = (struct cuda_ctx *) ctx;
+    if (this->thread_status == NOT_THR_CURRENT) {
+        this->thread_status = THR_CURRENT;
+        LOCKDBG("Locking.\n");
+        _papi_hwi_lock(_cuda_pw_lock);
+        LOCKDBG("Locked.\n");
+    }
+    if (control->cuctx_arr == NULL) {
+        res = cupti_init_cuctx_arr(&(control->cuctx_arr));
```

Comment:
I would not use this naming here. The upper layer does not need to know what is being initialised. Can you use `cupti_ctl_attr_init` instead? Also have `control->cupti_attr` instead of `cuctx_arr`.

Reply:
Variable names are intended for developers. Changing these names would only obscure the purpose of this function from developers. `cuctx_arr` is already a `void *`, at least the name should reflect its true purpose. The upper layer *must* call in to the lower layer to initialize the array of cuda contexts, it should be called that. Cuda contexts are confusing as it is, we should keep it clear here.

Reply:
Following this logic the framework layer should know exactly what is in the control state and context of the thread. That is not the case. The reason for having abstractions is to separate concerns.

Reply:
By renaming the function and the structure the abstraction won't change, only the meaning of what that statement does.

Reply:
My comment was aesthetic. Change `cupti_init_ctx_arr()` to `cupti_attr_init()` and `cuctx_arr` to `cupti_attr`. That way these are attributes of some sort. You are not giving away it is an array of cuda context to the upper layer.